File: pipeline.js
var Pipeline = {};
(function() {
  var aggregate = Pipeline.aggregate = function(actions) {
    var numResults = 0;
    var results = [];
    var gotError = false;
    function handleResults(callbacks) {
      numResults++;
      if (numResults === actions.length) {
        if (gotError) {
          callbacks.error();
        } else {
          callbacks.success(results);
        }
      }
    };

    return function(callbacks) {
      each(actions, function(action) {
        action({
          success: function(newResults) {
            results.push(newResults);
            handleResults(callbacks);
          },
          error: function() {
            gotError = true;
            handleResults(callbacks);
          }
        });
      });
    };
  };

  var converter = Pipeline.converter = function(conversion) {
    return function(action) {
      return function(callbacks) {
        action({
          success: function() {
            var result = conversion.apply(undefined, arguments);
            if (result && result.length && result.length > 1 && result.length == callbacks.success.length) {
              callbacks.success.apply(undefined, result);
            } else {
              callbacks.success(result);
            }
          },
          error: callbacks.error
        });
      };
    };
  };

  var concatenate = Pipeline.concatenate = converter(function(results) {
    return [].concat.apply([], results);
  });

  var group = Pipeline.group = function() {
    var actions = arguments;
    var numResults = 0;
    var results = [];
    var gotError = false;
    function handleResults(callbacks) {
      numResults++;
      if (numResults === actions.length) {
        if (gotError) {
          callbacks.error();
        } else {
          callbacks.success.apply(undefined, results);
        }
      }
    };

    return function(callbacks) {
      each(actions, function(action, index) {
        action({
          success: function(newResults) {
            results[index] = newResults;
            handleResults(callbacks);
          },
          error: function() {
            gotError = true;
            handleResults(callbacks);
          }
        });
      });
    };
  };
});

File: pipeline.hs
type Callback a = a -> IO()
type Action a = Callback a -> IO()

mkFetch :: a -> String -> Action a
mkFetch result address callback = remote address >>= passTo callback
    where remote _ = return result

passTo :: Callback a -> a -> IO()
passTo callback results = callback results >> return ()

fetch :: String -> Action [Int]
fetch = mkFetch [1, 3]

fetch' :: String -> Action String
fetch' = mkFetch "prefix"

converter :: (a -> b) -> Action a -> Action b
converter conversion action callback =  action (callback . conversion)

aggregator :: [Action a] -> Action [a]
aggregator actions callback = f actions []
 where f [] results = passTo callback results
       f (action:actions) results = action (\result -> f actions (result:results))

concatenator :: Action [[a]] -> Action [a]
concatenator = converter concat

sequencer :: Action a -> (a -> Action b) -> Action b
sequencer first second callback = first (\result -> (second result) callback)

application = fetchAll operate

operate :: String -> IO()
operate = print

fetchData = (converter show) . (converter sum) . concatenator . aggregator $ [fetch "first", fetch "second"]

fetchData' = fetch' "other"

fetchAll = (converter join) . aggregator $ [fetchData, fetchData']

join xs = xs!!0 ++ xs!!1
